### PR TITLE
Added function to add user to edwAdmin

### PR DESCRIPTION
**Pull request by hc-jordenlowe**
Added SQL script to Add-UserOrGroupToEdwAdmin.  Double check my SQL and see if it sucks

Added function Add-AccountToEDWAdmin to add a user to EDWAdmin

Added function call right below adding the user to DosAdmin

I need to double check if the user exists to not add them

### CODE REVIEW

File: Fabric.Authorization.API/scripts/Install-Authorization-Windows.ps1 [hunk 1]
```diff
@@ -147,6 +147,36 @@ function Add-User($authUrl, $name, $accessToken) {
     return Invoke-Post $url $body $accessToken
 }
 
+function Add-UserOrGroupToEdwAdmin($userOrGroup, $connString) {
+	$query = "DECLARE @userId INTEGER;
+			  DECLARE @roleId INTEGER;
+			  IF NOT EXISTS (SELECT * FROM [EDWAdmin].[CatalystAdmin].[IdentityBASE] WHERE [IdentityNM] = @identityName)
```
Comment (michaelvidal24) [inline Fabric.Authorization.API/scripts/Install-Authorization-Windows.ps1]: Do we need to handle adding the user if the user doesn't already exist in IdentityBASE?

File: Fabric.Authorization.API/scripts/Install-Authorization-Utilities.psm1 [hunk 2]
```diff
@@ -385,6 +411,27 @@ function Add-ListOfUsersToDosAdminGroup($edwAdminUsers, $connString, $authorizat
...
+function Add-AccountToEDWAdmin($accountName, $domain, $connString) {
+	$samAccountName = Get-SamAccountFromAccountName -accountName $accountName
+	if ((Test-IsUser -samAccountName $samAccountName -domain $domain) -or (Test-IsGroup -samAccountName $samAccountName -domain $domain)) {
+		try {
+			
+			Add-UserOrGroupToEdwAdmin -userOrGroup $accountName -connString $connString
+		}
+		catch {
+            $exception = $_.Exception
+            if ($exception.Response -ne $null) {
```
Comment (michaelvidal24) [inline Fabric.Authorization.API/scripts/Install-Authorization-Utilities.psm1]: The exception in this case will never have a Response object, that only exists on System.Net.WebException I believe, which won't be thrown here.

File: Fabric.Authorization.API/scripts/Install-Authorization-Windows.ps1 (hunk 1 above)
Comment (hc-jordenlowe) [reply Fabric.Authorization.API/scripts/Install-Authorization-Windows.ps1]: thats what this line is for:

IF NOT EXISTS (SELECT * FROM [EDWAdmin].[CatalystAdmin].[IdentityBASE] WHERE [IdentityNM] = @identityName)

however, if the user exists and they do not have permission, then I need to handle this case

File: Fabric.Authorization.API/scripts/Install-Authorization-Utilities.psm1 (hunk 2 above)
Comment (hc-jordenlowe) [reply Fabric.Authorization.API/scripts/Install-Authorization-Utilities.psm1]: Ah, good catch.  I was wondering why the exceptions didnt make sense

Comment (hc-jordenlowe) [reply Fabric.Authorization.API/scripts/Install-Authorization-Utilities.psm1]: Removed the exception altogether.  No need to catch it.

It was a case of copy pasta

File: Fabric.Authorization.API/scripts/Install-Authorization-Windows.ps1 (hunk 1 above)
Comment (hc-jordenlowe) [reply Fabric.Authorization.API/scripts/Install-Authorization-Windows.ps1]: Tested this functionality.  It will insert the user if its not there, then it will check to see if that user has access and will add it if they do not.

File: Fabric.Authorization.API/scripts/Install-Authorization-Utilities.psm1 [hunk 3]
```diff
@@ -270,6 +270,43 @@ function Add-RoleToGroup
     Invoke-Sql $connString $query @{groupId = $groupId; roleId = $roleId; ; clientId = $clientId} | Out-Null
 }
 
+function Add-UserOrGroupToEdwAdmin($userOrGroup, $connString) {
+	$query   = "DECLARE @userId INTEGER;
+                DECLARE @roleId INTEGER;
+                
+                SET @userId = -1
+                SELECT @userId = COALESCE([IdentityID], -1) FROM [EDWAdmin].[CatalystAdmin].[IdentityBASE] WHERE [IdentityNM] = @identityName
```
Comment (hckenmiller) [inline Fabric.Authorization.API/scripts/Install-Authorization-Utilities.psm1]: this is a global variable, right?

File: Fabric.Authorization.API/scripts/Install-Authorization-Utilities.psm1 [hunk 4]
```diff
@@ -270,6 +270,43 @@ function Add-RoleToGroup
...
+                        INSERT INTO [EDWAdmin].[CatalystAdmin].[IdentityBASE] ([IdentityNM])
+                        VALUES (@identityName);
+                        SELECT @userId = SCOPE_IDENTITY();
+                    END
+                
+                    SELECT @roleId = [RoleID]
+                    FROM [EDWAdmin].[CatalystAdmin].[RoleBASE]
+                    WHERE [RoleNM] = @roleName;
+                
+                    SELECT @userId, @roleId;
```
Comment (hckenmiller) [inline Fabric.Authorization.API/scripts/Install-Authorization-Utilities.psm1]: This Select is probably only to test the values, remove.

File: Fabric.Authorization.API/scripts/Install-Authorization-Windows.ps1 [hunk 5]
```diff
@@ -147,6 +147,44 @@ function Add-User($authUrl, $name, $accessToken) {
     return Invoke-Post $url $body $accessToken
 }
 
+function Add-UserOrGroupToEdwAdmin($userOrGroup, $connString) {
+	$query   = "DECLARE @userId INTEGER;
```
Comment (ryanorbaker) [inline Fabric.Authorization.API/scripts/Install-Authorization-Windows.ps1]: why is this here as well as utilities?  
I must be missing something?

File: Fabric.Authorization.API/scripts/Install-Authorization-Utilities.psm1 [hunk 6]
```diff
@@ -270,6 +270,43 @@ function Add-RoleToGroup
...
+                    IF (@userId < 0)
+                    BEGIN
+                        INSERT INTO [EDWAdmin].[CatalystAdmin].[IdentityBASE] ([IdentityNM])
+                        VALUES (@identityName);
+                        SELECT @userId = SCOPE_IDENTITY();
+                    END
+                
+                    SELECT @roleId = [RoleID]
+                    FROM [EDWAdmin].[CatalystAdmin].[RoleBASE]
+                    WHERE [RoleNM] = @roleName;
```
Comment (hckenmiller) [inline Fabric.Authorization.API/scripts/Install-Authorization-Utilities.psm1]: global?

File: Fabric.Authorization.API/scripts/Install-Authorization-Utilities.psm1 [hunk 7]
```diff
@@ -270,6 +270,43 @@ function Add-RoleToGroup
...
+                    END
+                        
+                    COMMIT;
+                END TRY
+                BEGIN CATCH
+                    ROLLBACK;
+                    THROW; 
+                END CATCH"
+                
+    Invoke-Sql -connectionString $connString -sql $query -parameters @{roleName = "EDW Admin"; identityName = $userOrGroup} | Out-Null
```
Comment (hckenmiller) [inline Fabric.Authorization.API/scripts/Install-Authorization-Utilities.psm1]: global? userorgroup

File: Fabric.Authorization.API/scripts/Install-Authorization-Utilities.tests.ps1 [hunk 8]
```diff
@@ -1443,3 +1443,66 @@ Describe 'Add-AccountToDosAdminGroup tests' -Tag 'Unit'{
...
+            Mock Test-IsGroup { return $false }
+            Mock Write-DosMessage {}
+
+            # Act and Assert
+            { Add-AccountToEDWAdmin -accountName "test\user" -domain "domain" -connString "connString" } | Should -Throw
+        }
+
+        It 'Should throw error if domain\user is incorrect' {
+            # Arrange
+            Mock Test-IsUser { return $True }
```
Comment (hckenmiller) [inline Fabric.Authorization.API/scripts/Install-Authorization-Utilities.tests.ps1]: lowercase to match all the others $true

File: Fabric.Authorization.API/scripts/Install-Authorization-Windows.ps1 [hunk 9]
```diff
@@ -147,6 +147,44 @@ function Add-User($authUrl, $name, $accessToken) {
...
+                        INSERT INTO [EDWAdmin].[CatalystAdmin].[IdentityBASE] ([IdentityNM])
+                        VALUES (@identityName);
+                        SELECT @userId = SCOPE_IDENTITY();
+                    END
+                
+                    SELECT @roleId = [RoleID]
+                    FROM [EDWAdmin].[CatalystAdmin].[RoleBASE]
+                    WHERE [RoleNM] = @roleName;
+                
+                    SELECT @userId, @roleId;
```
Comment (hckenmiller) [inline Fabric.Authorization.API/scripts/Install-Authorization-Windows.ps1]: remove it only is to test the values

File: Fabric.Authorization.API/scripts/Install-Authorization-Utilities.psm1 (hunk 3 above)
Comment (hc-jordenlowe) [reply Fabric.Authorization.API/scripts/Install-Authorization-Utilities.psm1]: That is a sql parameter variable.

File: Fabric.Authorization.API/scripts/Install-Authorization-Windows.ps1 (hunk 5 above)
Comment (hc-jordenlowe) [reply Fabric.Authorization.API/scripts/Install-Authorization-Windows.ps1]: So if you look, there are two files with the same functions.  That is because one is a Utilities powershell script and the other is this, which is the older powershell functions.  Anyways, as I understand the older powershell script, this one, is depreciated and will be removed eventually and replaced by the utility script and the quiet install.

So if it is going to be removed soon, do we want to reuse the code or just let the old code die?

File: Fabric.Authorization.API/scripts/Install-Authorization-Utilities.psm1 (hunk 6 above)
Comment (hc-jordenlowe) [reply Fabric.Authorization.API/scripts/Install-Authorization-Utilities.psm1]: This is a sql parameter name

File: Fabric.Authorization.API/scripts/Install-Authorization-Utilities.psm1 (hunk 3 above)
Comment (hckenmiller) [reply Fabric.Authorization.API/scripts/Install-Authorization-Utilities.psm1]: I'm not seeing where it gets it's value set

File: Fabric.Authorization.API/scripts/Install-Authorization-Windows.ps1 (hunk 5 above)
Comment (michaelvidal24) [reply Fabric.Authorization.API/scripts/Install-Authorization-Windows.ps1]: After this change, I think we can let the old code die. No need to do additional work to include the utilities module here.

File: Fabric.Authorization.API/scripts/Install-Authorization-Utilities.psm1 [hunk 10]
```diff
@@ -270,6 +270,41 @@ function Add-RoleToGroup
...
```
Comment (hckenmiller) [inline Fabric.Authorization.API/scripts/Install-Authorization-Utilities.psm1]: This is where roleName and identity are assigned.

File: Fabric.Authorization.API/scripts/Install-Authorization-Utilities.psm1 (hunk 6 above)
Comment (hckenmiller) [reply Fabric.Authorization.API/scripts/Install-Authorization-Utilities.psm1]: Yah I figure it out by looking at the invoke where the -parameters are declared/defined.
